### PR TITLE
Added before dependents to Modules

### PR DESCRIPTION
**Pull request by gregw**
Allow a module to list modules that it will be ordered before

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

**Comment by gregw** [comment]
@joakime @sbordet regardless of how we name this mod section, it is important to add it soon.  Having the ability to insert a module before another would be very handy for issuing patches to existing releases (eg for CVE fixes).

File: documentation/jetty-documentation/src/main/asciidoc/operations-guide/modules/modules.adoc [hunk 1]
```diff
@@ -198,16 +198,50 @@ Circular dependencies are not allowed.
...
-===== [optional]
+[[og-modules-directive-after]]
+===== [after]
 
-A list of module names that this module depends on, if they are explicitly enabled.
+This directive indicates that this module is ordered after the listed module names, if they are enabled.
 
-For example, the module `https` has an optional dependency on `http2`.
+For example, module `https` is `[after]` module `http2`.
 Enabling the `https` modules _does not_ enable the `http2` module.
```
Comment (janbartel) [inline documentation/jetty-documentation/src/main/asciidoc/operations-guide/modules/modules.adoc]:
```suggestion
Enabling the `https` module _does not_ enable the `http2` module.
```

File: documentation/jetty-documentation/src/main/asciidoc/operations-guide/modules/modules.adoc [hunk 2]
```diff
@@ -198,16 +198,50 @@ Circular dependencies are not allowed.
...
 
-However, if the `http2` module is explicitly enabled, then the `https` module depends on it, and therefore the `http2` module is xref:og-modules-directive-depends[sorted] _before_ the `https` module.
-In this way, you are guaranteed that the `http2` module is processed before the `https` module.
+However, if the `http2` module is enabled (explicitly or transitively), then the `https` module is xref:og-modules-directive-depends[sorted] _after_ the `http2` module.
```
Comment (janbartel) [inline documentation/jetty-documentation/src/main/asciidoc/operations-guide/modules/modules.adoc:209]:
```suggestion
However, if the `http2` module is then subsequently enabled (explicitly or transitively), then the `https` module is xref:og-modules-directive-depends[sorted] _after_ the `http2` module.
```

File: documentation/jetty-documentation/src/main/asciidoc/operations-guide/modules/modules.adoc [hunk 3]
```diff
@@ -198,16 +198,50 @@ Circular dependencies are not allowed.
...
+In this way, you are guaranteed that the `https` module is processed after the `http2` module.
+
+[[og-modules-directive-before]]
+===== [before]
+
+This directive indicates that this module is ordered before the listed module names, if they are enabled.
+
+For example, in order to create a custom `org.eclipse.jetty.server.Server` subclass instance, to which the standard configuration is applied, without modifying existing module files or XML files, you can create a `custom-server` xref:og-modules-custom[Jetty custom module]:
```
Comment (janbartel) [inline documentation/jetty-documentation/src/main/asciidoc/operations-guide/modules/modules.adoc]:
```suggestion
One use of this ordering could be to ensure that an xml configuration file is executed before any that are associated with the modules listed in the `before` clause. For example: in order to create a custom `org.eclipse.jetty.server.Server` subclass instance without needing to modify existing module or XML files, and to which the standard configuration is applied,  you can create a `custom-server` xref:og-modules-custom[Jetty custom module]:
```

File: jetty-start/src/main/java/org/eclipse/jetty/start/Module.java [hunk 4]
```diff
@@ -509,9 +521,24 @@ public String toString()
         return new HashSet<>(_provides);
     }
 
+    public Set<String> getBefore()
+    {
+        return Set.copyOf(_before);
```
Comment (janbartel) [inline jetty-start/src/main/java/org/eclipse/jetty/start/Module.java:526]: Probably would have been more efficient to return an immutable view of the Set, but as this was previous behaviour to return a copy ....

**Review by janbartel — CHANGES_REQUESTED** [review]

File: documentation/jetty-documentation/src/main/asciidoc/operations-guide/modules/modules.adoc (hunk 2 above)
Comment (gregw) [reply documentation/jetty-documentation/src/main/asciidoc/operations-guide/modules/modules.adoc:209]: It is not subsequently.  It may already be enabled.

Comment (janbartel) [reply documentation/jetty-documentation/src/main/asciidoc/operations-guide/modules/modules.adoc:209]: OK, yes.

**Review by sbordet — APPROVED** [review]

**Review by janbartel — APPROVED** [review]

**Pull request merged**